File: phases/p3/closer/app/routes/outreach.py
"""REST API routes for email outreach."""

from datetime import datetime, timezone

from fastapi import APIRouter, HTTPException, Query

from ..email_generator import generate_email
from ..email_sender import send_email
from ..preview import render_preview
from ..rate_limiter import rate_limiter
from ..queue import email_queue, QueuedEmail
from ..logger import outreach_logger
from ..export import export_logs
from ..delivery_tracker import track_delivery
from ..config import settings

router = APIRouter()


@router.post("/generate")
async def generate_outreach(
    application_id: str,
    job: dict | None = None,
    application: dict | None = None,
    recipient_name: str = "",
    recipient_email: str = "",
    template_type: str = "cold",
    company_name: str = "",
    job_title: str = "",
):
    """Generate an email draft for a given application."""
    try:
        email = generate_email(
            template_type=template_type,
            job=job,
            application=application,
            recipient_name=recipient_name,
            recipient_email=recipient_email,
        )
        # Extract company/job info from params (fall back to dict if not passed directly)
        if not company_name:
            company_name = (job or {}).get("company", "") or (application or {}).get("company", "")
        if not job_title:
            job_title = (job or {}).get("title", "") or (application or {}).get("role", "") or ""
        # Log the draft
        log_id = outreach_logger.log_send(
            application_id=application_id,
            user_id="demo-user",
            recipient_email=recipient_email,
            recipient_name=recipient_name,
            subject=email["subject"],
            status="draft",
            body_html=email["body_html"],
            body_text=email["body_text"],
            company_name=company_name,
            job_title=job_title,
        )
        # Inject tracking pixel into HTML body
        pixel_url = f"{settings.base_url}/api/outreach/{log_id}/track-open"
        tracked_html = email["body_html"] + (
            f'<img src="{pixel_url}" width="1" height="1" alt="" style="display:none;" />'
        )
        # Update stored body_html to include tracking pixel
        outreach_logger.update_log_status(
            log_id, status="draft", extras={"body_html": tracked_html}
        )
        return {
            "draft": {
                "id": log_id,
                "subject": email["subject"],
                "body_html": tracked_html,
                "body_text": email["body_text"],
            },
            "applicationId": application_id,
        }
    except FileNotFoundError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{outreach_id}/preview")
async def preview_outreach(outreach_id: str):
    """Preview a generated email."""
    # In production, fetch the email from DB by ID
    preview = render_preview()
    return preview


@router.post("/{outreach_id}/send")
async def send_outreach(
    outreach_id: str,
    user_id: str = "demo-user",
    subject: str = "",
    body_html: str = "",
    body_text: str = "",
    to_email: str = "",
    to_name: str = "",
    application_id: str = "",
):
    """Send an email. Respects dry-run, rate limits, and queuing."""
    # If body content not provided, try to look it up from the stored log
    log_entry = outreach_logger.get_log_entry(outreach_id)
    if log_entry:
        if log_entry.get("body_html"):
            body_html = body_html or log_entry["body_html"]
        if log_entry.get("body_text"):
            body_text = body_text or log_entry["body_text"]
        if log_entry.get("recipient_email"):
            to_email = to_email or log_entry["recipient_email"]
        if log_entry.get("recipient_name"):
            to_name = to_name or log_entry["recipient_name"]
        if log_entry.get("subject"):
            subject = subject or log_entry["subject"]
        if log_entry.get("application_id"):
            application_id = application_id or log_entry["application_id"]

    # Rate limit check
    allowed, retry_after = rate_limiter.check(user_id)
    if not allowed:
        raise HTTPException(
            status_code=429,
            detail=f"Rate limit exceeded. Retry after {retry_after}s",
            headers={"Retry-After": str(retry_after)},
        )

    # Check dry-run mode
    if settings.dry_run or settings.send_mode == "draft":
        outreach_logger.update_log_status(
            outreach_id,
            status="draft",
            extras={
                "recipient_email": to_email,
                "recipient_name": to_name,
                "subject": subject,
                "body_html": body_html,
                "body_text": body_text,
            },
        )
        return {
            "status": "draft",
            "message": "Draft updated (dry-run mode)",
            "logId": outreach_id,
        }

    # Send via SMTP
    result = send_email(
        to_email=to_email,
        to_name=to_name,
        subject=subject,
        body_html=body_html,
        body_text=body_text,
    )

    if not result.success:
        return {"status": "failed", "error": result.error}

    # Update the original draft's status to sent (keeps same ID)
    outreach_logger.update_log_status(
        outreach_id,
        status="sent",
        extras={
            "recipient_email": to_email,
            "recipient_name": to_name,
            "subject": subject,
            "body_html": body_html,
            "body_text": body_text,
        },
    )

    return {
        "status": "sent",
        "messageId": result.message_id,
        "logId": outreach_id,
        "sentAt": datetime.now(timezone.utc).isoformat(),
    }


@router.post("/{outreach_id}/queue")
async def queue_outreach(
    outreach_id: str,
    to_email: str = "",
    to_name: str = "",
    subject: str = "",
    body_html: str = "",
    body_text: str = "",
):
    """Queue an email for delayed sending."""
    email = QueuedEmail(
        id=outreach_id,
        to_email=to_email,
        to_name=to_name,
        subject=subject,
        body_html=body_html,
        body_text=body_text,
    )
    email_queue.enqueue(email)
    return {"status": "queued", "position": email_queue.size()}


@router.get("/logs")
async def get_logs(
    user_id: str = "demo-user",
    status: str | None = None,
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
):
    """Get outreach logs for a user."""
    logs = outreach_logger.get_logs(
        user_id=user_id, status=status, limit=limit, offset=offset
    )
    return {"logs": logs, "total": len(logs)}


@router.get("/stats")
async def get_stats(user_id: str = "demo-user"):
    """Get aggregated outreach stats."""
    stats = outreach_logger.get_stats(user_id)
    remaining = rate_limiter.remaining(user_id)
    return {**stats, **remaining}


@router.post("/{outreach_id}/track")
async def track_event(outreach_id: str, event: str, error: str | None = None):
    """Track a delivery event (bounce, open, reply, etc.)."""
    success = track_delivery(outreach_id, event, error)
    if not success:
        raise HTTPException(status_code=404, detail="Log entry not found")
    return {"status": "tracked", "event": event}


@router.get("/{outreach_id}/track-open")
async def track_open_pixel(outreach_id: str):
    """Tracking pixel endpoint — logs 'opened' event and returns a 1×1 transparent GIF."""
    track_delivery(outreach_id, "opened")
    # 1×1 transparent GIF
    gif = (
        b"GIF89a\x01\x00\x01\x00\x80\x00\x00"
        b"\xff\xff\xff\x00\x00\x00!\xf9\x04\x00"
        b"\x00\x00\x00\x00,\x00\x00\x00\x00\x01\x00"
        b"\x01\x00\x00\x02\x02D\x01\x00;"
    )
    from fastapi.responses import Response
    return Response(content=gif, media_type="image/gif")


@router.post("/{outreach_id}/cancel")
async def cancel_outreach(outreach_id: str):
    """Cancel / remove a draft from the queue."""
    success = outreach_logger.update_log_status(outreach_id, status="cancelled")
    if not success:
        raise HTTPException(status_code=404, detail="Log entry not found")
    return {"status": "cancelled", "logId": outreach_id}


@router.get("/export")
async def export_logs_csv(user_id: str = "demo-user"):
    """Export outreach logs as CSV."""
    csv = export_logs(user_id)
    from fastapi.responses import PlainTextResponse
    return PlainTextResponse(
        content=csv,
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=outreach_logs.csv"},
    )

File: phases/p0/apps/web/lib/outreach.ts
export interface OutreachQueueItem {
  id: string;
  companyName: string;
  jobTitle: string;
  recipientEmail: string;
  recipientName: string;
  subject: string;
  status: "draft" | "queued" | "sent" | "failed";
  createdAt: string;
  errorMessage?: string;
}

export interface OutreachDeliveryEntry {
  id: string;
  companyName: string;
  jobTitle: string;
  recipientEmail: string;
  recipientName: string;
  subject: string;
  sentAt: string;
  status: string;
  deliveryStatus: string;
  errorMessage?: string;
}

export interface OutreachLogResponse {
  id: string;
  application_id: string;
  user_id: string;
  recipient_email: string;
  recipient_name: string;
  company_name: string;
  job_title: string;
  subject: string;
  subject_hash: string;
  status: string;
  delivery_status: string;
  error_message: string | null;
  body_html: string;
  body_text: string;
  sent_at: string;
  opened_at: string | null;
  replied_at: string | null;
}

export interface OutreachStatsResponse {
  total: number;
  sent: number;
  opened: number;
  replied: number;
  bounced: number;
  failed: number;
  open_rate: number;
  reply_rate: number;
  bounce_rate: number;
  hourly_remaining: number;
  daily_remaining: number;
}

interface OutreachLogsResponse {
  logs: OutreachLogResponse[];
  total: number;
}

export function toQueueItem(log: OutreachLogResponse): OutreachQueueItem {
  return {
    id: log.id,
    companyName: log.company_name || "",
    jobTitle: log.job_title || "",
    recipientEmail: log.recipient_email || "",
    recipientName: log.recipient_name || "",
    subject: log.subject,
    status: mapQueueStatus(log),
    createdAt: log.sent_at,
    errorMessage: log.error_message || undefined,
  };
}

export function toDeliveryEntry(log: OutreachLogResponse): OutreachDeliveryEntry {
  return {
    id: log.id,
    companyName: log.company_name || "",
    jobTitle: log.job_title || "",
    recipientEmail: log.recipient_email || "",
    recipientName: log.recipient_name || "",
    subject: log.subject,
    sentAt: log.sent_at,
    status: log.status,
    deliveryStatus: log.delivery_status || log.status,
    errorMessage: log.error_message || undefined,
  };
}

export async function getOutreachStats(): Promise<OutreachStatsResponse> {
  return fetchJson<OutreachStatsResponse>("/api/outreach/stats");
}

export async function getOutreachLogs(options?: {
  status?: string;
  limit?: number;
  offset?: number;
}): Promise<OutreachLogResponse[]> {
  const params = new URLSearchParams();
  if (options?.status) {
    params.set("status", options.status);
  }
  if (typeof options?.limit === "number") {
    params.set("limit", String(options.limit));
  }
  if (typeof options?.offset === "number") {
    params.set("offset", String(options.offset));
  }

  const query = params.toString();
  const data = await fetchJson<OutreachLogsResponse>(
    `/api/outreach/logs${query ? `?${query}` : ""}`,
  );
  return data.logs ?? [];
}

function mapQueueStatus(log: OutreachLogResponse): OutreachQueueItem["status"] {
  if (log.status === "draft") {
    return "draft";
  }
  if (log.status === "failed") {
    return "failed";
  }
  if (log.delivery_status === "pending") {
    return "queued";
  }
  return "sent";
}

function formatRecipient(log: OutreachLogResponse): string {
  if (log.recipient_name && log.recipient_email) {
    return `${log.recipient_name} <${log.recipient_email}>`;
  }
  return log.recipient_name || log.recipient_email || "Unknown recipient";
}

export async function sendOutreach(
  id: string,
  overrides?: { toEmail?: string; toName?: string; subject?: string },
): Promise<{ status: string; messageId?: string; logId?: string; sentAt?: string; error?: string }> {
  const body: Record<string, string> = {};
  if (overrides?.toEmail) body.to_email = overrides.toEmail;
  if (overrides?.toName) body.to_name = overrides.toName;
  if (overrides?.subject) body.subject = overrides.subject;

  const res = await fetch(`/api/outreach/${id}/send`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    let message = res.statusText || "Failed to send";
    try {
      const payload = (await res.json()) as { error?: string; detail?: string };
      message = payload.error ?? payload.detail ?? message;
    } catch {
      // Ignore non-JSON bodies
    }
    throw new Error(message);
  }
  return res.json() as Promise<{ status: string; messageId?: string; logId?: string; sentAt?: string; error?: string }>;
}

export async function trackOutreach(
  id: string,
  event: string,
  error?: string,
): Promise<{ status: string; event: string }> {
  const params = new URLSearchParams({ event });
  if (error) params.set("error", error);
  const res = await fetch(`/api/outreach/${id}/track?${params.toString()}`, { method: "POST" });
  if (!res.ok) {
    let message = res.statusText || "Failed to track";
    try {
      const payload = (await res.json()) as { error?: string; detail?: string };
      message = payload.error ?? payload.detail ?? message;
    } catch {
      // Ignore non-JSON bodies
    }
    throw new Error(message);
  }
  return res.json() as Promise<{ status: string; event: string }>;
}

export async function cancelOutreach(id: string): Promise<{ status: string; logId?: string }> {
  const res = await fetch(`/api/outreach/${id}/cancel`, { method: "POST" });
  if (!res.ok) {
    let message = res.statusText || "Failed to cancel";
    try {
      const payload = (await res.json()) as { error?: string; detail?: string };
      message = payload.error ?? payload.detail ?? message;
    } catch {
      // Ignore non-JSON bodies
    }
    throw new Error(message);
  }
  return res.json() as Promise<{ status: string; logId?: string }>;
}

export async function generateOutreach(params: {
  applicationId: string;
  recipientName?: string;
  recipientEmail?: string;
  templateType?: string;
  job?: Record<string, unknown>;
  application?: Record<string, unknown>;
}): Promise<{ draft: { id: string; subject: string; body_html: string; body_text: string }; applicationId: string }> {
  const query = new URLSearchParams();
  query.set("application_id", params.applicationId);
  if (params.recipientName) query.set("recipient_name", params.recipientName);
  if (params.recipientEmail) query.set("recipient_email", params.recipientEmail);
  if (params.templateType) query.set("template_type", params.templateType);
  if (params.job) {
    query.set("job", JSON.stringify(params.job));
    query.set("company_name", String(params.job.company ?? ""));
    query.set("job_title", String(params.job.title ?? ""));
  }
  if (params.application) {
    query.set("application", JSON.stringify(params.application));
    if (!params.job) {
      query.set("company_name", String((params.application as Record<string, unknown>).company ?? ""));
      query.set("job_title", String((params.application as Record<string, unknown>).role ?? ""));
    }
  }

  const res = await fetch(`/api/outreach/generate?${query.toString()}`, { method: "POST" });
  if (!res.ok) {
    let message = res.statusText || "Failed to generate";
    try {
      const payload = (await res.json()) as { error?: string; detail?: string };
      message = payload.error ?? payload.detail ?? message;
    } catch {
      // Ignore non-JSON bodies
    }
    throw new Error(message);
  }
  return res.json() as Promise<{ draft: { id: string; subject: string; body_html: string; body_text: string }; applicationId: string }>;
}

async function fetchJson<T>(input: RequestInfo | URL): Promise<T> {
  const response = await fetch(input, { cache: "no-store" });
  if (!response.ok) {
    let message = response.statusText || "Request failed";
    try {
      const payload = (await response.json()) as { error?: string; detail?: string };
      message = payload.error ?? payload.detail ?? message;
    } catch {
      // Ignore non-JSON error bodies.
    }
    throw new Error(message);
  }

  try {
    return (await response.json()) as T;
  } catch {
    throw new Error("Unexpected response from server");
  }
}
